File: shared/contracts/package_manifest.py
"""Tur-55 C9: Package Manifest contract (Python).

JSON Schema: package_manifest.schema.json
Bu dataclass hem validasyon hem de serileştirme sağlar.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class PackageManifest:
    name: str = ""
    version: str = ""
    arch: str = ""
    size_bytes: int = 0
    sha256: str = ""
    dependencies: list[str] = field(default_factory=list)
    signed: bool = False
    signature_path: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "version": self.version,
            "arch": self.arch,
            "size_bytes": self.size_bytes,
            "sha256": self.sha256,
            "dependencies": list(self.dependencies),
            "signed": self.signed,
            "signature_path": self.signature_path,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> PackageManifest:
        return cls(
            name=data.get("name", ""),
            version=data.get("version", ""),
            arch=data.get("arch", ""),
            size_bytes=int(data.get("size_bytes", 0)),
            sha256=data.get("sha256", ""),
            dependencies=list(data.get("dependencies", [])),
            signed=bool(data.get("signed", False)),
            signature_path=data.get("signature_path", ""),
        )

File: shared/contracts/package_manifest.ts
// Tur-55 C9: Package Manifest contract (TypeScript mirror).
// Python `PackageManifest` dataclass'ına karşılık gelen TS arayüzü.
// JSON Schema: ../../shared/contracts/package_manifest.schema.json
export interface PackageManifest {
  name: string;
  version: string;
  arch: "amd64" | "arm64" | "i386" | "any";
  size_bytes: number;
  sha256: string; // 64-hex
  dependencies: string[];
  signed: boolean;
  signature_path?: string;
}

export function isPackageManifest(obj: unknown): obj is PackageManifest {
  if (typeof obj !== "object" || obj === null) return false;
  const m = obj as Record<string, unknown>;
  return (
    typeof m.name === "string" &&
    typeof m.version === "string" &&
    typeof m.arch === "string" &&
    typeof m.size_bytes === "number" &&
    typeof m.sha256 === "string" &&
    Array.isArray(m.dependencies)
  );
}
